ratis-server/src/main/java/org/apache/ratis/server/impl/ReadIndexHeartbeats.java

```diff
@@ -0,0 +1,168 @@
+
+    AppendEntriesListener(long commitIndex) {
+      this.commitIndex = commitIndex;
+    }
+
+    CompletableFuture<Long> getFuture() {
+      return future;
+    }
+
+    boolean receive(LogAppender logAppender, AppendEntriesReplyProto proto,
+                    Predicate<Predicate<RaftPeerId>> hasMajority) {
+      if (isCompletedNormally()) {
+        return true;
+      }
+
+      final HeartbeatAck reply = replies.computeIfAbsent(
+          logAppender.getFollowerId(), key -> new HeartbeatAck(logAppender));
+      if (reply.receive(proto)) {
+        if (hasMajority.test(id -> replies.get(id).isAcknowledged())) {
+          future.complete(commitIndex);
+          return true;
+        }
+      }
+
+      return isCompletedNormally();
+    }
+
+    boolean isCompletedNormally() {
+      return future.isDone() && !future.isCancelled() && !future.isCompletedExceptionally();
+    }
+  }
+
+  class AppendEntriesListeners {
+    private final NavigableMap<Long, AppendEntriesListener> sorted = new TreeMap<>();
+
+    synchronized AppendEntriesListener add(long commitIndex, Function<Long, AppendEntriesListener> constructor) {
+      return sorted.computeIfAbsent(commitIndex, constructor);
+    }
+
+    synchronized void onAppendEntriesReply(LogAppender appender, AppendEntriesReplyProto reply,
+                                           Predicate<Predicate<RaftPeerId>> hasMajority) {
+      final long callId = reply.getServerReply().getCallId();
+
+      Iterator<Map.Entry<Long, AppendEntriesListener>> iterator = sorted.entrySet().iterator();
+      while (iterator.hasNext()) {
+        Map.Entry<Long, AppendEntriesListener> entry = iterator.next();
+        if (entry.getKey() > callId) {
+          return;
+        }
+
+        final AppendEntriesListener listener = entry.getValue();
+        if (listener == null) {
+          continue;
+        }
+
+        if (listener.receive(appender, reply, hasMajority)) {
+          ackedCommitIndex.updateToMax(listener.commitIndex, s -> LOG.debug("{}: {}", this, s));
+          iterator.remove();
+        }
+      }
+    }
+  }
+
+  private final AppendEntriesListeners appendEntriesListeners = new AppendEntriesListeners();
+  private final RaftLogIndex ackedCommitIndex = new RaftLogIndex("ackedCommitIndex", RaftLog.INVALID_LOG_INDEX);
+
+  AppendEntriesListener addAppendEntriesListener(long commitIndex,
+                                                              Function<Long, AppendEntriesListener> constructor) {
+    if (commitIndex <= ackedCommitIndex.get()) {
+      return null;
+    }
+    return appendEntriesListeners.add(commitIndex, constructor);
+  }
+
+  void onAppendEntriesReply(LogAppender appender, AppendEntriesReplyProto reply,
+                            Predicate<Predicate<RaftPeerId>> hasMajority) {
+    appendEntriesListeners.onAppendEntriesReply(appender, reply, hasMajority);
+  }
+
+  void clear() {
+    appendEntriesListeners.sorted.clear();
+  }
```

Comment:
We should complete all the futures with NotLeaderException; see `WatchRequests.failWatches(..)`.

Reply:
done